File: cpp_files/raylib/shape_survivors/src/bullets/bullet.cpp
#include "../systems.hpp"
#include "bullet.hpp"

static const float BULLET_RADIUS = 4.0f;

Bullet::Bullet(const BulletConfig& config, const Vector2& start_pos, const Vector2& target_pos)
    : position({start_pos.x, start_pos.y}),

      circle({.radius = BULLET_RADIUS}),

      color({.tint = RED}),

      movement({.speed = config.speed, .direction = {0, 0}}),

      damage({.current = config.damage}),

      lifetime({.max = config.lifetime, .current = config.lifetime}) {
    //
    Systems::set_direction(position, movement, target_pos);
}

void Bullet::update(const float dt) {
    Systems::move(position, movement, dt);
    Systems::update_lifetime(lifetime, dt);
}

void Bullet::draw(void) const {
    Systems::draw_circle(position, circle, color);
}

File: python_files/raylib_files/shape_survivors/bullet.py
from dataclasses import dataclass

import pyray as p

import components
import systems


@dataclass(slots=True)
class BulletConfig:
    damage: int
    speed: float
    lifetime: float


BULLET_RADIUS: float = 4.0


class Bullet:
    def __init__(
        self, config: BulletConfig, start_pos: p.Vector2, target_pos: p.Vector2
    ) -> None:
        self.position: components.Position = components.Position(
            start_pos.x, start_pos.y
        )

        self.circle: components.Circle = components.Circle(radius=BULLET_RADIUS)

        self.color: components.Color = components.Color(tint=p.RED)

        self.movement: components.Movement = components.Movement(
            speed=config.speed,
            direction=p.Vector2(0, 0),
        )

        self.damage: components.Damage = components.Damage(current=config.damage)

        self.lifetime: components.Lifetime = components.Lifetime(
            max=config.lifetime, current=config.lifetime
        )

        systems.set_direction(self.position, self.movement, target_pos)

    def update(self, dt: float) -> None:
        systems.move(self.position, self.movement, dt)
        systems.update_lifetime(self.lifetime, dt)

    def draw(self) -> None:
        systems.draw_circle(self.position, self.circle, self.color)
